processor/samplingprocessor/tailsamplingprocessor/processor.go

```diff
@@ -166,33 +166,44 @@ func (tsp *tailSamplingSpanProcessor) samplingPolicyOnTick() {
 			}
 
 			trace.Decisions[i] = decision
+		}
 
-			switch decision {
-			case sampling.Sampled:
-				stats.RecordWithTags(
-					policy.ctx,
-					[]tag.Mutator{tag.Insert(tagSampledKey, "true")},
-					statCountTracesSampled.M(int64(1)),
-				)
-				decisionSampled++
-
-				trace.Lock()
-				traceBatches := trace.ReceivedBatches
-				trace.Unlock()
-
-				for j := 0; j < len(traceBatches); j++ {
-					tsp.nextConsumer.ConsumeTraces(policy.ctx, internaldata.OCToTraceData(traceBatches[j]))
-				}
-			case sampling.NotSampled:
-				stats.RecordWithTags(
-					policy.ctx,
-					[]tag.Mutator{tag.Insert(tagSampledKey, "false")},
-					statCountTracesSampled.M(int64(1)),
-				)
-				decisionNotSampled++
+		finalDecision := sampling.NotSampled
```

Comment:
Can you please add some comments to explain what the code is doing? Even better would be to break this function into smaller ones, it is very large and difficult to understand.

Reply:
I agree this code is difficult to understand. Instead of trying to fix/refactor the existing implementation do you think its worthwhile to rewrite this processor to rely on the new groupbytrace? The implementation would be much simpler.

I asked @jpkrohling about this and he was open to the idea. I can create an issue to discuss further if you'd like.

Reply:
@chris-smith-zocdoc are you suggesting to discard this PR and instead submit a bigger refactoring? I am open to the idea, but it depends on what exactly is the refactoring.

Reply:
I'm going to open a separate issue for it. With the deadlock [bug](https://github.com/open-telemetry/opentelemetry-collector/issues/1811) I found its not something we could do yet anyway.